cmd/ocp-build-data-enforcer/main.go

```diff
@@ -0,0 +1,250 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/openshift/imagebuilder"
+	dockercmd "github.com/openshift/imagebuilder/dockerfile/command"
+	"github.com/pmezard/go-difflib/difflib"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openshift/ci-tools/pkg/github"
+)
+
+type options struct {
+	ocpBuildDataRepoDir string
+}
+
+func gatherOptions() *options {
+	o := &options{}
+	flag.StringVar(&o.ocpBuildDataRepoDir, "ocp-build-data-repo-dir", "../ocp-build-data", "The directory in which the ocp-build-data reposity is")
+	flag.Parse()
+	return o
+}
+func main() {
+	opts := gatherOptions()
+
+	configs, err := gatherAllOCPImageConfigs(opts.ocpBuildDataRepoDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to gather all ocp image configs")
+	}
+
+	streamMap, err := readStreamMap(opts.ocpBuildDataRepoDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to read streamMap")
+	}
+
+	for cfgIdx := range configs {
+		dereferenceStreams(&configs[cfgIdx], streamMap)
+	}
+
+	errGroup := &errgroup.Group{}
+	for idx := range configs {
+		errGroup.Go(func() error {
+			processDockerfile(configs[idx])
+			return nil
+		})
+	}
+	if err := errGroup.Wait(); err != nil {
+		logrus.WithError(err).Fatal("Processing failed")
+	}
+	logrus.Infof("Processed %d configs", len(configs))
+}
+
+func dereferenceStreams(config *ocpImageConfig, streamMap streamMap) {
+	if replacement, hasReplacement := streamMap[config.From.Stream]; hasReplacement && replacement.UpstreamImage != nil {
+		config.From.Stream = *replacement.UpstreamImage
+	}
+	for blder := range config.From.Builder {
+		if replacement, hasReplacement := streamMap[config.From.Builder[blder].Stream]; hasReplacement && replacement.UpstreamImage != nil {
+			config.From.Builder[blder].Stream = *replacement.UpstreamImage
+		}
+	}
+}
+
+func processDockerfile(config ocpImageConfig) {
+	orgRepo := config.orgRepo()
+	log := logrus.WithField("file", config.SourceFileName).WithField("org/repo", orgRepo)
+	split := strings.Split(orgRepo, "/")
+	if n := len(split); n != 2 {
+		log.Errorf("splitting orgRepo didn't yield 2 but %d results", n)
+		return
+	}
+	if split[0] == "openshift-priv" {
+		log.Trace("Ignoring repo in openshift-priv org")
+		return
+	}
+	org, repo := split[0], split[1]
+	getter := github.FileGetterFactory(org, repo, "release-4.6")
+
+	log = log.WithField("dockerfile", config.dockerfile())
+	data, err := getter(config.dockerfile())
+	if err != nil {
+		log.WithError(err).Error("Failed to get dockerfile")
+		return
+	}
+	if len(data) == 0 {
+		log.Error("dockerfile is empty")
```

Comment:
Sorry, re-opening because I see a couple examples where the message isn't accurate.
- Distgit only repositories
I see in the logs this error being reported for `images/logging-eventrouter.yml` (openshift/ose-logging-eventrouter). The reason you can't find this Dockerfile is because it has no `content` field in the ocp-build-data metadata. This means that this is a "dist-git" only build (nothing sourced from upstream). You can safely ignore these metadata files since there is nothing upstream to reconcile.

- Aliased repositories
`images/openshift-enterprise-tests.yml` has a `content:` field, but it references an `alias` instead of a `url`. Aliases must be resolved through group.yaml's [sources](https://github.com/openshift/ocp-build-data/blob/17c591b2d53fc70577f14127ed2816f9406a4dc9/group.yml#L69-L73). 

- True broken repos
Other images were broken at the time (e.g. `images/jenkins-agent-nodejs-12-rhel7.yml`) because not all PRs to enable rhel8 have landed. Running the same code now, I would not expect to see `images/jenkins-agent-nodejs-12-rhel7.yml` because https://github.com/openshift/jenkins/blob/master/agent-nodejs-12/Dockerfile.rhel8 is now available.